addons/hr_contract/models/hr_contract.py

```diff
@@ -109,6 +104,27 @@ def _onchange_employee_id(self):
             self.department_id = self.employee_id.department_id
             self.resource_calendar_id = self.employee_id.resource_calendar_id
 
+    @api.constrains('employee_id', 'state', 'date_start', 'date_end')
+    def _check_current_contract(self):
+        """ Two contracts in state [incoming | pending | open] cannot overlap """
+        for contract in self.filtered(lambda c: c.state not in ['draft', 'close', 'cancel']):
+            domain = [
+                ('id', '!=', contract.id),
+                ('employee_id', '=', contract.employee_id.id),
+                ('state', 'in', ['incoming', 'pending', 'open']),
+            ]
+
+            if not contract.date_end:
+                start_domain = [('date_start', '>=', contract.date_start)]
```

Comment:
In this case I think the final domain should look like this `['|', ('date_end', '>=', contract.date_start), ('date_end', '=', False)]`
Which means 
```
start_domain = []
end_domain = ['|', ('date_end', '>=', contract.date_start), ('date_end', '=', False)]
```